src/robot/running/keywordrunner.py

```diff
@@ -206,16 +256,99 @@ def _to_number_with_arithmetics(self, item):
             raise TypeError("Expected number, got %s." % type_name(item))
         return number
 
-    def _run_one_round(self, data, values):
-        name = ', '.join(format_assign_message(var, item)
-                         for var, item in zip(data.variables, values))
-        result = KeywordResult(kwname=name,
-                               type=data.FOR_ITEM_TYPE)
-        for var, value in zip(data.variables, values):
-            self._context.variables[var] = value
-        runner = KeywordRunner(self._context, self._templated)
-        with StatusReporter(self._context, result):
-            runner.run_keywords(data.keywords)
+
+class ForInZipLoopRunner(BasicForLoopRunner):
+    def __init__(self, context, templated=False):
+        super(ForInZipLoopRunner, self).__init__(context, templated)
+
+    def _flavor_name(self):
+        return 'IN ZIP'
+
+    def _replace_variables(self, data):
+        """
+        Make sure we have as many loop-variables as zipped-iterables.
+
+        (Other than that, delegate to the superclass.)
+        """
+        values = super(ForInZipLoopRunner, self)._replace_variables(data)
+        if len(data.variables) == len(data.values):
+            return values
+        raise DataError("Expected %d Loop variables, but found %d"
+                % (len(data.values), len(data.variables)))
+
+    def _transform_items(self, items):
+        answer = list()
+        for item in items:
+            if not is_list_like(item):
+                # Checking here because _validate() doesn't have access
+                # to the actual values.
+                raise DataError("For-In-Zip Loop items must all be List-like (but not Strings); got %s with value '%s'" % (item.__class__, item))
```

Comment:
I'll try to rephrase the comment to better reflect this point: Usually one would expect validation to happen in _validate(), but this particular validation relies on information that is not available to _validate() so instead it's here.